tests/ui/needless_deref.rs

```diff
@@ -0,0 +1,26 @@
+fn main() {}
+
+mod should_lint {
+    fn foo() {
+        let a = &12;
+        let _b = &*a;
+
+        let s = &String::new();
+        let a: &str = &*s;
+    }
+}
+
+mod should_not_lint {
+    fn foo(x: &mut &u32) {
+        *x = &1;
+    }
+
+    fn main() {
+        let mut x = &0;
+        foo(&mut &*x); // should not lint
```

Comment:
What's wrong with removing `&*` here?

Reply:
```
let mut x = &0;
foo(&mut &*x); 
// foo(&mut x); // assert will fail
// foo(&mut &**x); // compile fail
assert_eq!(*x, 0);
```
So `foo(&mut &*x);` is the only correct way.
Whenever `&*(&T)` is wrapped with `&` or `&mut`, the lint should ignore it.

Reply:
Very interesting.

> Whenever `&*(&T)` is wrapped with `&` or `&mut`, the lint should ignore it.

Wouldn't it only be when wrapped with `&mut`? An immutable borrow is no risk since it's immutable I think.

Could you also add that the `&T` must be a `ExprKind::Path` where the base is a mutable binding? For example `&mut &*fun()` should be lintable.

Reply:
## Why the lint excludes `& &* (&T)` case
```
fn main() {
    let x = &12;
    let addr_x = &x as *const _ as usize;
    let addr_y= &&*x as *const _ as usize; // assert ok
    // let addr_y = &x as *const _ as usize; // assert fail                                                                                       
    assert_ne!(addr_x, addr_y);
}
```

Reply:
> Could you also add that the &T must be a ExprKind::Path where the base is a mutable binding? For example &mut &*fun() should be lintable.

Sure, I have done it.

Reply:
I think if we suppressed lints whenever it would change the pointer equality of two expressions, that would be too limiting.

Reply:
I'll second @camsteffen on pointer equality here. It's very unlikely to be used for that purpose. I might consider allowing `let x = &*y` as an explicit reborrow.

Reply:
Don't forget to handle `&mut &*x.y`.

Reply:
> I might consider allowing let x = &*y as an explicit reborrow.

If any others think so? Or I'll retract this pr.

Reply:
> > I might consider allowing let x = &*y as an explicit reborrow.
> 
> If any others think so? Or I'll retract this pr.

This seems like a misunderstanding. @Jarcho is suggesting to allow a `let` statement that that doesn't do anything except "re-borrow". That doesn't apply to most instances of this lint.

Reply:
@camsteffen @Jarcho No longer lint on reborrow now.
Frankly speaking, my intuitive is to lint on reborrow.

Reply:
I just realized I had a typo earlier. I meant to say consider allowing the lint on `let x = &&*y`. This would be used for when a different pointer value is actually needed. I'm not sure if there's ever a need for it, but it is a semantic change and this gives the person a more explicit way to do that without having to allow the lint.

As for `let x = &*y` I had erroneously mentioned earlier, I think it only matters if `y` is a mutable reference, as then it's changing the mutability of the pointer which shouldn't be linted.

Reply:
> As for let x = &*y I had erroneously mentioned earlier, I think it only matters if y is a mutable reference, as then it's changing the mutability of the reference which shouldn't be linted.

This pr doesn't deal with `&mut T`.

Reply:
@Jarcho Thanks. I guess we have the same view now!

@camsteffen What do you think of `let x = &&*y` ? I guess Jarcho and I support not linting on it.
In `cargo linkcheck`, I didn't find such pattern.

Reply:
The `let` cases seem a bit speculative to me. Personally I would "keep it simple" until a real FP is found.

Reply:
Is everything resolved here?